xarray/core/parallelcompat.py

```diff
@@ -0,0 +1,271 @@
+    """
+
+    # TODO this list is probably redundant with something inside xarray.apply_ufunc
+    ALLOWED_NON_CHUNKED_TYPES = {int, float, np.ndarray}
+
+    chunked_arrays = [
+        a
+        for a in args
+        if is_chunked_array(a) and type(a) not in ALLOWED_NON_CHUNKED_TYPES
+    ]
+
+    # Asserts all arrays are the same type (or numpy etc.)
+    chunked_array_types = {type(a) for a in chunked_arrays}
+    if len(chunked_array_types) > 1:
+        raise TypeError(
+            f"Mixing chunked array types is not supported, but received multiple types: {chunked_array_types}"
+        )
+    elif len(chunked_array_types) == 0:
+        raise TypeError("Expected a chunked array but none were found")
+
+    # iterate over defined chunk managers, seeing if each recognises this array type
+    chunked_arr = chunked_arrays[0]
+    chunkmanagers = list_chunkmanagers()
+    for chunkmanager in chunkmanagers.values():
+        if chunkmanager.is_chunked_array(chunked_arr):
+            return chunkmanager
+
+    raise TypeError(
+        f"Could not find a Chunk Manager which recognises type {type(chunked_arr)}"
+    )
+
+
+class ChunkManagerEntrypoint(ABC, Generic[T_ChunkedArray]):
+    """
+    Adapter between a particular parallel computing framework and xarray.
+
+    Attributes
+    ----------
+    array_cls
+        Type of the array class this parallel computing framework provides.
+
+        Parallel frameworks need to provide an array class that supports the array API standard.
+        Used for type checking.
+    """
+
+    array_cls: type[T_ChunkedArray]
+    available: bool = True
+
+    @abstractmethod
+    def __init__(self):
+        ...
+
+    def is_chunked_array(self, data: Any) -> bool:
+        return isinstance(data, self.array_cls)
+
+    @abstractmethod
+    def chunks(self, data: T_ChunkedArray) -> T_NormalizedChunks:
+        ...
+
+    @abstractmethod
+    def normalize_chunks(
+        self,
+        chunks: T_Chunks,
+        shape: tuple[int] | None = None,
+        limit: int | None = None,
+        dtype: np.dtype | None = None,
+        previous_chunks: T_NormalizedChunks | None = None,
+    ) -> T_NormalizedChunks:
+        """Called by open_dataset"""
+        ...
+
+    @abstractmethod
+    def from_array(
+        self, data: np.ndarray, chunks: T_Chunks, **kwargs
+    ) -> T_ChunkedArray:
+        """Called when .chunk is called on an xarray object that is not already chunked."""
+        ...
+
+    def rechunk(
+        self,
+        data: T_ChunkedArray,
+        chunks: T_NormalizedChunks | tuple[int, ...] | T_Chunks,
+        **kwargs,
+    ) -> T_ChunkedArray:
+        """Called when .chunk is called on an xarray object that is already chunked."""
+        return data.rechunk(chunks, **kwargs)  # type: ignore[attr-defined]
+
+    @abstractmethod
+    def compute(self, data: T_ChunkedArray, **kwargs) -> np.ndarray:
+        ...
```

Comment:
This is an abstract method - surely it shouldn't have any implementation, even raising? I mean that code would never possibly be called.

I tried to make some of the methods on the `ChunkManagerEntrypoint` ABC abstract and some non-abstract but `NotImplemented` because I wanted to give some indication of which methods were necessary for the chunked backend to work at all, and which were only necessary to implement in order for some subset of xarray's to work.

Reply:
I guess on the other hand there is no harm in adding `raise NotImplementedError()`, and you're right it does make the code a little clearer.

Reply:
note that because `compute` is marked as a `abstractmethod`, trying to instantiate a subclass that does not override it will cause the constructor to fail, so technically this does not make any difference at all because it will never be executed, and we could also use `pass` instead (which I think is the canonical way to say "empty block", but I'm not sure)